twenty-first/src/util_types/algebraic_hasher.rs

```diff
@@ -48,26 +48,48 @@ pub trait AlgebraicHasher: Clone + Send + Sync {
         Self::sample_index(seed, (1 << 16) * other::roundup_npo2(max as u64) as usize) % max
     }
 
-    /// Given a uniform random `seed` digest, a `max` that is a power of two,
-    /// produce `count` uniform random numbers (sample indices) in the interval
-    /// `[0; max)`. The seed should be a Fiat-Shamir digest to ensure a high
+    /// Given a uniform random `seed` digest, a `max_index` that is a power of two,
+    /// produce `num_indices` uniform random numbers (sample indices) in the interval
+    /// `[0; max_index)`. The seed should be a Fiat-Shamir digest to ensure a high
     /// degree of randomness.
     ///
-    /// - `count`: The number of sample indices
-    /// - `seed`: A hash digest
-    /// - `max`: The (non-inclusive) upper bound (a power of two)
-    fn sample_indices(count: usize, seed: &Digest, max: usize) -> Vec<usize> {
-        Self::get_n_hash_rounds(seed, count)
+    /// - `seed`: A hash `Digest`
+    /// - `max_index`: The (non-inclusive) upper bound (a power of two)
+    /// - `num_indices`: The number of sample indices
+    fn sample_indices(seed: &Digest, max_index: usize, num_indices: usize) -> Vec<usize> {
+        Self::get_n_hash_rounds(seed, num_indices)
             .iter()
-            .map(|random_input| Self::sample_index(random_input, max))
+            .map(|random_input| Self::sample_index(random_input, max_index))
+            .collect()
+    }
+
+    /// Given a uniform random `seed` digest, produce `num_weights` uniform random
+    /// `XFieldElement`s (sample weights). The seed should be a Fiat-Shamir digest
+    /// to ensure a high degree of randomness.
+    ///
+    /// - `seed`: A hash `Digest`
+    /// - `num_weights`: The number of sample weights
+    fn sample_weights(seed: &Digest, num_weights: usize) -> Vec<XFieldElement> {
+        Self::get_n_hash_rounds(seed, num_weights)
+            .iter()
+            .map(XFieldElement::sample)
             .collect()
     }
 
     fn get_n_hash_rounds(seed: &Digest, count: usize) -> Vec<Digest> {
         let mut digests = Vec::with_capacity(count);
         (0..count)
             .into_par_iter()
-            .map(|i: usize| Self::hash_slice(&[seed.to_sequence(), i.to_sequence()].concat()))
+            .map(|counter: usize| {
+                let counter = Digest::new([
+                    BFieldElement::new(counter as u64),
+                    BFieldElement::zero(),
+                    BFieldElement::zero(),
+                    BFieldElement::zero(),
+                    BFieldElement::zero(),
+                ]);
+                Self::hash_pair(&counter, seed)
+            })
```

Comment:
@aszepieniec, could I get your input here?

Reply:
Technically, this only allows $2^{64} - 2^{32}$ samples, after which it starts repeating.

A more general solution could be to encode the counter as a `U32s<N>`.

Considering we need `security_level` (160) samples in the practical case, the current should be sufficient.

I've added the following assertion:

```rust
assert!(count <= BFieldElement::MAX as usize);
```